README.adoc
= Home Assistant add-ons repository

image::https://img.shields.io/github/license/fboulay/ha-addons-repository.svg[licence]

== About

Here, you will find all my add-ons for Home Assistant

== Installation

Adding this add-ons repository to your Home Assistant instance is
pretty easy. Follow https://home-assistant.io/hassio/installing_third_party_addons[the official instructions] on the
website of Home Assistant, and use the following URL: 

```
https://github.com/fboulay/ha-addons-repository
```

== Available add-ons

=== link:ha-protonmail-bridge/DOCS.md[ProtonMail Bridge]

* Install the ProtonMail bridge to send emails using your ProtonMail account


==== About ProtonMail bridge

ProtonMail Bridge is an application available to all paid users that enables the integration of your ProtonMail account with any SMTP clients (in theory). 
Bridge runs in the background by seamlessly encrypting messages as they leave Home Assistant.

This addon allow to send notification emails using a ProtonMail account. 

== License

MIT License

Copyright (c) 2021 Florian Boulay

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


ha-protonmail-bridge/CHANGELOG.md
# Changelog

## 1.4.0

### 🚀 Improvements

- Add translations for the configuration options. Available languages
  are french and english.

### 🐛 Bug fixes

- Fix the docker build. The go dependency was not available anymore.
- The addon is working fine again. A custom git repository is used to 
  make it work on ARM architectures.

### ✅ Misc updates

- Config file is now is yaml format.

## 1.3.4

THIS ADDON DOES NOT WORK ANYMORE, read carefully the explanation below.

### ✅ Misc

- The ProtonMail server does not support anymore the ProtonMail Bridge 
  with the version 1.4.5. Thus this addon does not work anymore. I am
  working on a new working version. For now an error message is 
  displayed in the log file.

## 1.3.3

### 🚀 Improvements

- Update base docker image to 3.14
- Improve security rating to 6
  
### 🐛 Bug fixes

- Fix typo in logs
  
## 1.3.2

### 📚 Documentation

- Instructions on how to add the addon repository were wrong. The
  URL is now fixed.
  
## 1.3.1

### 🐛 Bug fixes

- Fix Dockerfile. The multi-stage build was not working as expected
  for building multi-arch images

### 📚 Documentation

- Improved and fixed documentation

## 1.3.0

### 🚀 Improvements

- The addon is now running using S6 overlay.
- Check in `run.sh` if all mandatory parameters are present.
- Set versions explicitly in Dockerfile for packages installed with `apk`
- Add hadolint check in Github Action

### 🐛 Bug fixes

- Github Action to update Docker Hub description is now working

## 1.2.2

### 🚀 Improvements

- Github action now update the Docker Hub description when pushing
  a new image.
- Add Shellcheck on every push.

## 1.2.1

### 🐛 Bug fixes

- Github action to publish to Docker Hub is now working

## 1.2.0

### 🚀 Improvements

- Add a security check on the ProtonMail password against the HaveIBeenPwned
  database. If the password is not safe, a warning is displayed in the logs.
- Support amd64 architecture
- Automatically push docker image to Docker Hub when a PR is merged on main

### 🐛 Bug fixes

- The addon can sometimes start to late and the SMTP integration does not manage
  to communicate with it. Changed the startup option to `system`.

### 📚 Documentation

- Add badges about supported architecture
- Add badge about supported licence
- Add labels in the Dockerfile
  
## 1.1.0

### 🚀 Improvements

- Expects scripts are now wrapped in their own files
- `run.sh` entrypoint is more clever, and now calls only needed scripts
- ProtonMail Bridge output is hidden. Only relevant information is
   displayed. This is thanks to expect scripts.
- Add a `debug.sh` file which is used only when developing to interact
  with expect scripts.
- Logs are using utility functions from bashio. The date format displays
  the current date and has a millisecond resolution.

## 1.0.0

- First working version
  
## 0.0.1

- First non working version

## Categories

### 🚀 Improvements

### 🐛 Bug fixes

### 📚 Documentation

### 🠝 Dependency updates

### ✅ Misc updates


ha-protonmail-bridge/debug.sh
#!/usr/bin/env bashio

#
# debug.sh : this can be used to have an interactive menu and call all expect scripts in any order to debug them
#

# Create directories we want to store in persistent directory
mkdir -p /data/.password-store
mkdir -p /data/.gnupg 2>/dev/null
mkdir -p /data/.cache 2>/dev/null
mkdir -p /data/.config 2>/dev/null

chmod 700 /data/.gnupg
chmod 700 /data/.password-store

# socat will make the conn appear to come from 127.0.0.1
# ProtonMail Bridge currently expects that.
# It also allows us to bind to the real ports :)
socat TCP-LISTEN:25,fork TCP:127.0.0.1:1025 &
socat TCP-LISTEN:143,fork TCP:127.0.0.1:1143 &

# A new one is created each time the container is starting. It does not seem to create any issue
gpg --no-options --generate-key --batch /protonmail/gpgparams
pass init pass-key


set +o errexit 

while :
do
    echo ""
    echo "Menu"
    echo " i : info of current connected account"
    echo " c : connect to ProtonMail"
    echo " l : logout from ProtonMail"
    echo " e : exit this menu"
    read -r -p ">" c
	case $c in
        i) RET=$(/src/info.expect) 
            echo $?
            echo "$RET"
            ;;
        c) RET=$(/src/add-account.expect  "$(bashio::config 'username')" "$(bashio::config 'password')" "$(bashio::config 'two_factor_code')") 
            echo $?
            echo "$RET"
            ;;
        l) RET=$(/src/logout.expect) 
            echo $?
            echo "$RET"
            ;;
        e) break
            ;;
        *) echo "Select between i, c, l and e only"
	esac
done


ha-protonmail-bridge/run.sh
#!/usr/bin/env bashio

bashio::log.info "Staring ProtonMail Bridge addon"

bashio::config.require.username
bashio::config.require.password

# Create directories we want to store in persistent directory
mkdir -p /data/.password-store
mkdir -p /data/.gnupg 
mkdir -p /data/.cache 
mkdir -p /data/.config 

chmod 700 /data/.gnupg
chmod 700 /data/.password-store

# socat will make the conn appear to come from 127.0.0.1
# ProtonMail Bridge currently expects that.
# It also allows us to bind to the real ports :)
socat TCP-LISTEN:25,fork TCP:127.0.0.1:1025 &
socat TCP-LISTEN:143,fork TCP:127.0.0.1:1143 &

# Disable exit on non-zero status, because expect script can return non-zero values
set +o errexit 

# Generate a PGP key if missing

if ! gpg --list-keys pass-key &> /dev/null; then
    bashio::log.info "First time launch - Generating a PGP key"
    gpg --no-options --generate-key --batch --no-tty /protonmail/gpgparams &> /dev/null
    pass init pass-key &> /dev/null
fi

# Check if the password is in the HaveIBeenPwned database
bashio::config.suggest.safe_password 'password'

# A new one is created if not connected to ProtonMail
/src/info.expect  > /dev/null
if [[ $? == 1 ]]; then
    bashio::log.info "Not connected - adding account to ProtonMail Bridge"
    set +o errexit 
    
    # Display error on add account
    if ! PMB_ADD_ACCOUNT=$(/src/add-account.expect "$(bashio::config 'username')" "$(bashio::config 'password')" "$(bashio::config 'two_factor_code')"); then
        bashio::log.error "${PMB_ADD_ACCOUNT}" 
    fi
fi

set +o errexit 
PMB_INFO=$(/src/info.expect) 
if [[ $? == 2 ]]; then
    set -o errexit 
    bashio::log.info "" 
    bashio::log.info "${PMB_INFO}"
    bashio::log.info ""
    bashio::log.info "Starting ProtonMail Bridge" 
    /protonmail/proton-bridge --noninteractive
else
    bashio::log.error "Cannot connect - Please check errors above"
fi
